lib/rust/performance-logger/src/lib.rs

```diff
@@ -0,0 +1,395 @@
+/// an interval.
+fn start_interval_label(log_level: LogLevel, interval_name: &str) -> String {
+    format!(
+        "{1}{0}{2}{0}start",
+        MESSAGE_DELIMITER,
+        serde_plain::to_string(&log_level).unwrap(),
+        interval_name
+    )
+}
+
+/// Return a string that encodes the given log level and name for a mark that indicates the end of
+/// an interval.
+fn end_interval_label(log_level: LogLevel, interval_name: &str) -> String {
+    format!(
+        "{1}{0}{2}{0}end",
+        MESSAGE_DELIMITER,
+        serde_plain::to_string(&log_level).unwrap(),
+        interval_name
+    )
+}
+
+/// Return a string that encodes the given log level and name for a measurement.
+fn measure_interval_label(log_level: LogLevel, interval_name: &str) -> String {
+    format!(
+        "{1}{0}{2}{0}measure",
+        MESSAGE_DELIMITER,
+        serde_plain::to_string(&log_level).unwrap(),
+        interval_name
+    )
+}
+
+/// Start measuring an interval. Returns a `IntervalHandle` that an be used to end the created
+/// interval. The interval can also be ended by calling `end_interval` with the same label and log
+/// level.
+pub fn start_interval(log_level: LogLevel, label: &str) -> IntervalHandle {
+    let interval_name = ImString::from(label);
+    if log_level_is_active(log_level) {
+        performance().mark(&start_interval_label(log_level, &interval_name)).unwrap();
+    }
+    IntervalHandle::new(log_level, &interval_name)
+}
+
+/// End measuring an interval. Return the measurement taken between start and end of the interval,
+/// if possible.
+pub fn end_interval(
+    log_level: LogLevel,
+    interval_name: &str,
+) -> Result<Measurement, MeasurementError> {
+    let start_label = start_interval_label(log_level, interval_name);
+    let end_label = end_interval_label(log_level, interval_name);
+    let measurement_label = measure_interval_label(log_level, interval_name);
+    if !log_level_is_active(log_level) {
+        Err(MeasurementError::ProfilingDisabled)
+    } else {
+        performance().mark(&end_label).or(Err(MeasurementError::PerformanceAPIExecutionFailure))?;
+        performance()
+            .measure_with_start_mark_and_end_mark(&measurement_label, &start_label, &end_label)
+            .or(Err(MeasurementError::PerformanceAPIExecutionFailure))?;
+
+        let entries: js_sys::Array = performance().get_entries_by_type("measure");
+
+        if entries.length() < 1 {
+            return Err(MeasurementError::NoMeasurementFound);
+        }
+        let measure = entries.get(entries.length() - 1);
+        let measure: PerformanceEntry = measure.into();
+
+        let measurement: Measurement =
+            measure.try_into().or(Err(MeasurementError::InvalidMeasurementConversion))?;
+        Ok(measurement)
+    }
+}
+
+/// Handle that allows ending the interval.
+#[derive(Clone, Debug)]
+pub struct IntervalHandle {
+    interval_name: ImString,
+    log_level:     LogLevel,
+    released:      bool,
+}
```

Comment:
This is still in the section "profiling methods" but does not look like a method – please check if all sections are layout correctly.

Reply:
Changed to separate section.